core-backend/wasmtime/src/env.rs

```diff
@@ -18,269 +18,162 @@
 
 //! Wasmtime environment for running a module.
 
+use alloc::rc::Rc;
+
 use crate::memory::MemoryWrap;
 use alloc::{boxed::Box, collections::BTreeMap, format, string::ToString, vec::Vec};
 use gear_backend_common::{
-    funcs, BackendError, BackendReport, Environment, ExtInfo, TerminationReason,
+    funcs as common_funcs, BackendError, BackendReport, Environment, ExtInfo, TerminationReason,
 };
 use gear_core::{
     env::{Ext, LaterExt},
     memory::{Memory, PageBuf, PageNumber},
 };
 use wasmtime::{
-    Extern, Func, Instance, Limits, Memory as WasmtimeMemory, MemoryType, Module, Store, Trap,
+    Engine, Extern, Func, Instance, Memory as WasmtimeMemory, MemoryType, Module, Store, Trap,
 };
 
+pub struct LaterStore<T>(Rc<Store<T>>);
+
+impl<T> Clone for LaterStore<T> {
+    fn clone(&self) -> Self {
+        LaterStore(Rc::clone(&self.0))
+    }
+}
+
+impl<T: Default> LaterStore<T> {
+    pub fn new(eng: &Engine) -> Self {
+        LaterStore(Rc::from(Store::new(eng, T::default())))
+    }
+    /// In order to be able borrow mutable reference many times we need
+    /// to make it in unsafe manner.
+    /// Wasmtime store object must be mut borrowed to execute instance,
+    /// but also we must mut borrow it in memory sys-calls: alloc/free/...
+    /// But memory syscalls called in the same time with instance execution,
+    /// so there is no ways to avoid twice mut borrowing.
+    pub fn get_mut_ref(&mut self) -> &mut Store<T> {
+        unsafe {
+            let r = self.0.as_ref();
+            let ptr = r as *const Store<T> as *mut Store<T>;
+            ptr.as_mut().expect("ptr must be here")
```

Comment:
This is dangerous method. :bomb:  Do we have any alternatives?

Reply:
I can't find any other way. I don't understand why, but rust core/std completely restrict double mut borrowing. But we must do it. It can be done in C++, so must can be done in rust. Yes, it's unsafe, but only for programmer, who can make something wrong during double mut borrowing. But if we are careful then everything would be ok.

I must mention, that we may be can do VERY VERY big refactoring and avoid double mut borrowing, but first I'm not sure that  it can be done, second I think it can lead to some other big problems.

Reply:
Ok remove it